content/security/rate-limiting.md

```diff
@@ -1,34 +1,79 @@
+      }),
+    }),
+  ],
+})
+export class AppModule {}
 ```
 
-When there is a load balancer or reverse proxy between the server and the internet, Express may need to be configured to trust the headers set by the proxy in order to get the correct IP for the end user. To do so, first use the `NestExpressApplication` platform [interface](https://docs.nestjs.com/first-steps#platform) when creating your `app` instance, then enable the [trust proxy](https://expressjs.com/en/guide/behind-proxies.html) setting:
+The above will set the global options for the `ttl`, the time to live, and the `limit`, the maximum number of requests within the ttl, for the routes of your application that are guarded.
+
+Once the module has been imported, you can then choose how you would like to bind the `ThrottlerGuard`. Any kind of binding as mentioned in the [guards](https://docs.nestjs.com/guards) section is fine. If you wanted to bind the guard globally, for example, you could do so but adding this provider to any module
 
 ```typescript
-const app = await NestFactory.create<NestExpressApplication>(AppModule);
-// see https://expressjs.com/en/guide/behind-proxies.html
-app.set('trust proxy', 1);
+{
+  provide: APP_GUARD,
+  useClass: ThrottlerGuard
+}
 ```
 
-> info **Hint** If you use the `FastifyAdapter`, use the [fastify-rate-limit](https://github.com/fastify/fastify-rate-limit) package instead.
+#### Customization
+
+There may be a time where you want to bind the guard to a controller or globally, but want to avoid rate limiting one or more of your endpoints. For that, you can use the `@SkipThrottle()` decorator, to negate the throttler for an entire class or a single route. The `@SkipThrottle()` decorator can also take in a boolean for if there is a case where you want to exclude _most_ of a controller, but not every route.
+
+There is also the `@Throttle()` decorator which can be used to override the `limit` and `ttl` set in the global module, to give tighter or looser security options. This decorator can be used on a class or a function as well. The order for this decorator does matter, as the arguments are in the order of `limit` `ttl`.
+
+#### Websockets
+
+This module does work with websockets as well, with some limited functionality. First of all, user agent headers are not taken into consideration due to the difference in the underlying transport layer of Socket.IO vs Websockets. The other thing to make note of is that globally bound guards do not activate on websocket gateways, so you **must** bind the guard to the gateway itself using `@UseGuards()`.
+
+#### GraphQL
+
+Currently, only GraphQL with Express is supported, but Fastify support is coming as well. This module makes use of setting headers through the `res` object and reading headers through the `req` object of Express. To make sure these are available, when configuring your GraphQLModule, make sure the option `context: ({{ '{' }} req, res {{ '}' }}) => ({{ '{' }} req, res {{ '}' }})` is set.
```

Comment:
Btw `req` should be automatically set by the `@nestjs/graphql` package.

Reply:
`req` is there by default, but `res` needs to be added for this module, so I thought we should be explicit in adding it. [Though it's looking like Apollo adds `res` by default too](https://www.apollographql.com/docs/apollo-server/api/apollo-server/#middleware-specific-context-fields). Should we keep this, or make a note that this _should_ be default behavior?